editor/src/components/canvas/dom-walker.ts

```diff
@@ -212,44 +203,54 @@ export interface DomWalkerProps {
 }
 
-function mergeFragmentMetadata(
-  metadata: ReadonlyArray<ElementInstanceMetadata>,
-): ElementInstanceMetadataMap {
-  let working: MapLike<ElementInstanceMetadata> = {}
+// For performance reasons this function mutates the first parameter:
+// it adds elementMetadata to metadataToMutate in way that it merges fragments with the same elementpath
+// Note: it only mutates the map itself, it should never mutate the metadata instances inside the map!
+function mergeElementMetadataToMapWithFragments_MUTATE(
+  metadataToMutate: ElementInstanceMetadataMap,
+  elementMetadata: Readonly<ElementInstanceMetadata>,
+): void {
+  const pathString = EP.toString(elementMetadata.elementPath)
+  const existingMetadata = metadataToMutate[pathString]
 
-  fastForEach(metadata, (elementMetadata) => {
-    const pathString = EP.toString(elementMetadata.elementPath)
-    const existingMetadata = working[pathString]
+  if (existingMetadata == null) {
+    metadataToMutate[pathString] = elementMetadata
+  } else {
+    // We've hit a fragment, so remove the style etc., but keep the frames for selection
+    const merged = elementInstanceMetadata(
+      elementMetadata.elementPath,
+      left('fragment'),
+      boundingRectangle(
+        existingMetadata.globalFrame ?? zeroCanvasRect,
+        elementMetadata.globalFrame ?? zeroCanvasRect,
+      ),
+      boundingRectangle(
+        existingMetadata.localFrame ?? zeroLocalRect,
+        elementMetadata.localFrame ?? zeroLocalRect,
+      ),
+      false,
+      false,
+      emptySpecialSizeMeasurements,
+      {},
+      {},
+      null,
+      null,
+    )
 
-    if (existingMetadata == null) {
-      working[pathString] = elementMetadata
-    } else {
-      // We've hit a fragment, so remove the style etc., but keep the frames for selection
-      const merged = elementInstanceMetadata(
-        elementMetadata.elementPath,
-        left('fragment'),
-        boundingRectangle(
-          existingMetadata.globalFrame ?? zeroCanvasRect,
-          elementMetadata.globalFrame ?? zeroCanvasRect,
-        ),
-        boundingRectangle(
-          existingMetadata.localFrame ?? zeroLocalRect,
-          elementMetadata.localFrame ?? zeroLocalRect,
-        ),
-        false,
-        false,
-        emptySpecialSizeMeasurements,
-        {},
-        {},
-        null,
-        null,
-      )
+    metadataToMutate[pathString] = merged
+  }
+}
 
-      working[pathString] = merged
-    }
+// For performance reasons this function mutates the first parameter:
+// it merges otherMetadata into metadataToMutate in way that it merges fragments with the same elementpath
+// Note: it only mutates the map itself, it should never mutate the metadata instances inside the map!
+function mergeMetadataMapsWithFragments_MUTATE(
+  metadataToMutate: ElementInstanceMetadataMap,
+  otherMetadata: Readonly<ElementInstanceMetadataMap>,
+): void {
+  fastForEach(Object.values(otherMetadata), (elementMetadata) => {
```

Comment:
can this Object.values be slow here? genuine question!

Reply:
Well, I somehow have to iterate through the elements and `Object.values` is still way faster then `Object.keys`

See my benchmark:
<img width="1105" alt="image" src="https://user-images.githubusercontent.com/127662/173400420-85480e40-be50-44ee-9884-f96ee632a5ab.png">